presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -77,48 +82,100 @@ public static Expression rewriteGroupingOperation(GroupingOperation expression,
             );
-            resolveFunction(rewritten, functionArgumentTypes, metadata.getFunctionRegistry());
+            resolveFunction(rewritten, functionArgumentTypes, functionRegistry);
 
             return rewritten;
         }
     }
 
-    private static Expression translateFieldToLongLiteral(FieldId fieldId, RelationId requiredOriginRelationId)
+    private static int translateFieldToInteger(FieldId fieldId, RelationId requiredOriginRelationId)
     {
         // TODO: this section should be rewritten when support is added for GROUP BY columns to reference an outer scope
         checkState(fieldId.getRelationId().equals(requiredOriginRelationId), "grouping arguments must all come from the same relation");
-        return new LongLiteral(Integer.toString(fieldId.getFieldIndex()));
+        return fieldId.getFieldIndex();
+    }
+
+    /**
+     * The grouping function is used in conjunction with GROUPING SETS, ROLLUP and CUBE to
+     * indicate which columns are present in that grouping.
+     *
+     * <p>The grouping function must be invoked with arguments that exactly match the columns
+     * referenced in the corresponding GROUPING SET, ROLLUP or CUBE clause at the associated
+     * query level. Those column arguments are not evaluated and instead the function is
+     * re-written with the arguments below.
+     *
+     * <p>To compute the resulting bit set for a particular row, bits are assigned to the
+     * argument columns with the rightmost column being the most significant bit. For a
+     * given grouping, a bit is set to 0 if the corresponding column is included in the
+     * grouping and 1 otherwise. For an example, see the SQL documentation for the
+     * function.
+     *
+     * @param groupId An ordinal indicating which grouping is currently being processed.
+     *        Each grouping is assigned a unique monotonically increasing integer.
+     * @param columns The column arguments with which the function was
+     *        invoked converted to ordinals with respect to the base table column
+     *        ordering.
+     * @param groupingSetDescriptors A collection of ordinal lists where the index of
+     *        the list is the groupId and the list itself contains the ordinals of the
+     *        columns present in the grouping. For example: [[0, 2], [2], [0, 1, 2]]
+     *        means the the 0th list contains the set of columns that are present in
+     *        the 0th grouping.
+     * @return A bit set converted to decimal indicating which columns are present in
+     *         the grouping. If a column is NOT present in the grouping its corresponding
+     *         bit is set to 1 and to 0 if the column is present in the grouping.
+     */
+    static long calculateGrouping(long groupId, List<Integer> columns, List<List<Integer>> groupingSetDescriptors)
+    {
+        long grouping = (1L << columns.size()) - 1;
+
+        List<Integer> groupingSet = groupingSetDescriptors.get((int) groupId);
```

Comment:
Either use `int` for `groupId` or checked cast here.

Reply:
Done. Decided to use `Math.toIntExact()`.